MTD Validation: SIM clusters

casarsa opened the pull request:
#### PR description:

This PR builds the SIM clusters associated to the reconstructed BTL and ETL clusters and fills new monitoring histograms with:
- time, energy and position resolutions for BTL and ETL clusters;
- cluster time pulls vs energy and eta;
- Y vs X of the cluster local coordinates.

(@fabiocos , @parbol , @gsorrentino18 )

#### PR validation:

The new code has been validated with the TTbar WFs 34634.0 and 34834.0 with CMSSW_12_0_X_2021-06-21-1100.

--------------------------------------------------------------------------------------

**current code**

step 3 (WF 34634.0)

MemoryReport> Peak virtual size 6372.39 Mbytes
 Key events increasing vsize: 
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[1] run: 1 lumi: 1 event: 1  vsize = 6372.34 deltaVsize = 0 rss = 3253.47 delta = 0
[2] run: 1 lumi: 1 event: 2  vsize = 6372.39 deltaVsize = 0.046875 rss = 3338.56 delta = 85.0898
[8] run: 1 lumi: 1 event: 8  vsize = 6372.39 deltaVsize = 0.0078125 rss = 3397.37 delta = 58.8047
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[10] run: 1 lumi: 1 event: 10  vsize = 6372.39 deltaVsize = 0 rss = 3408.45 delta = 11.0859
[9] run: 1 lumi: 1 event: 9  vsize = 6372.39 deltaVsize = 0 rss = 3405.04 delta = 7.66797
[8] run: 1 lumi: 1 event: 8  vsize = 6372.39 deltaVsize = 0.0078125 rss = 3397.37 delta = 58.8047
TimeReport> Time report complete in 281.325 seconds
 Time Summary: 
 - Min event:   2.63794
 - Max event:   17.0326
 - Avg event:   5.85748
 - Total loop:  122.119
 - Total init:  154.509
 - Total job:   281.325
 - EventSetup Lock:   0
 - EventSetup Get:   0
 Event Throughput: 0.0818873 ev/s
 CPU Summary: 
 - Total loop:  115.4
 - Total init:  156.349
 - Total extra: 0
 - Total job:   275.98
 Processing Summary: 
 - Number of Events:  10
 - Number of Global Begin Lumi Calls:  1
 - Number of Global Begin Run Calls: 1

step 3 (WF 34834.0)

 MemoryReport> Peak virtual size 10830.3 Mbytes
 Key events increasing vsize: 
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[1] run: 1 lumi: 1 event: 1  vsize = 9805.86 deltaVsize = 0 rss = 7058.96 delta = 0
[2] run: 1 lumi: 1 event: 2  vsize = 10830.3 deltaVsize = 1024.49 rss = 7500.25 delta = 441.285
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[4] run: 1 lumi: 1 event: 4  vsize = 10830.3 deltaVsize = 0 rss = 7134.74 delta = -365.512
[3] run: 1 lumi: 1 event: 3  vsize = 10830.3 deltaVsize = 0 rss = 7416.19 delta = -84.0586
[2] run: 1 lumi: 1 event: 2  vsize = 10830.3 deltaVsize = 1024.49 rss = 7500.25 delta = 441.285
TimeReport> Time report complete in 2298.29 seconds
 Time Summary: 
 - Min event:   141.309
 - Max event:   332.148
 - Avg event:   186.191
 - Total loop:  2040.3
 - Total init:  252.703
 - Total job:   2298.29
 - EventSetup Lock:   0
 - EventSetup Get:   0
 Event Throughput: 0.00490124 ev/s
 CPU Summary: 
 - Total loop:  1494.69
 - Total init:  102.663
 - Total extra: 0
 - Total job:   1602.19
 Processing Summary: 
 - Number of Events:  10
 - Number of Global Begin Lumi Calls:  1
 - Number of Global Begin Run Calls: 1

--------------------------------------------------------------------------------------

**new code**

step 3 (WF 34634.0)

MemoryReport> Peak virtual size 6374.41 Mbytes
 Key events increasing vsize: 
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[1] run: 1 lumi: 1 event: 1  vsize = 6374.36 deltaVsize = 0 rss = 3375.52 delta = 0
[2] run: 1 lumi: 1 event: 2  vsize = 6374.41 deltaVsize = 0.046875 rss = 3448.57 delta = 73.0586
[8] run: 1 lumi: 1 event: 8  vsize = 6374.41 deltaVsize = 0.0078125 rss = 3500.94 delta = 52.3633
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[10] run: 1 lumi: 1 event: 10  vsize = 6374.41 deltaVsize = 0 rss = 3509.6 delta = 8.66016
[9] run: 1 lumi: 1 event: 9  vsize = 6374.41 deltaVsize = 0 rss = 3513.95 delta = 13.0078
[8] run: 1 lumi: 1 event: 8  vsize = 6374.41 deltaVsize = 0.0078125 rss = 3500.94 delta = 52.3633
TimeReport> Time report complete in 219.515 seconds
 Time Summary: 
 - Min event:   2.40397
 - Max event:   16.8583
 - Avg event:   5.15732
 - Total loop:  106.134
 - Total init:  109.03
 - Total job:   219.515
 - EventSetup Lock:   0
 - EventSetup Get:   0
 Event Throughput: 0.0942203 ev/s
 CPU Summary: 
 - Total loop:  100.611
 - Total init:  135.833
 - Total extra: 0
 - Total job:   240.449
 Processing Summary: 
 - Number of Events:  10
 - Number of Global Begin Lumi Calls:  1
 - Number of Global Begin Run Calls: 1

step 3 (WF 34834.0)

MemoryReport> Peak virtual size 10829.4 Mbytes
 Key events increasing vsize: 
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[1] run: 1 lumi: 1 event: 1  vsize = 9804.88 deltaVsize = 0 rss = 6334.34 delta = 0
[2] run: 1 lumi: 1 event: 2  vsize = 10829.4 deltaVsize = 1024.49 rss = 6906.25 delta = 571.91
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[4] run: 1 lumi: 1 event: 4  vsize = 10829.4 deltaVsize = 0 rss = 6600.7 delta = -305.551
[3] run: 1 lumi: 1 event: 3  vsize = 10829.4 deltaVsize = 0 rss = 6856.95 delta = -49.3008
[2] run: 1 lumi: 1 event: 2  vsize = 10829.4 deltaVsize = 1024.49 rss = 6906.25 delta = 571.91
TimeReport> Time report complete in 2217.3 seconds
 Time Summary: 
 - Min event:   143.84
 - Max event:   377.531
 - Avg event:   193.618
 - Total loop:  2092.58
 - Total init:  119.324
 - Total job:   2217.3
 - EventSetup Lock:   0
 - EventSetup Get:   0
 Event Throughput: 0.00477879 ev/s
 CPU Summary: 
 - Total loop:  1516.13
 - Total init:  96.4185
 - Total extra: 0
 - Total job:   1617.48
 Processing Summary: 
 - Number of Events:  10
 - Number of Global Begin Lumi Calls:  1
 - Number of Global Begin Run Calls: 1

--------------------------------------------------------------------------------------

**DQM file size**
current:
1129.74 KiB      MTD/ETL
239.63 KiB       MTD/BTL

new:
1535.33 KiB      MTD/ETL
560.66 KiB       MTD/BTL

casarsa committed:
RECO-SIM matching for BTL clusters

casarsa committed:
Fixed BTL clusters' local position

casarsa committed:
RECO-SIM matching for ETL clusters

casarsa committed:
Added validation plots with the SIM clusters

casarsa committed:
Added resolution plots for BTL clusters' rho and phi; applied minor fixes.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34207/23442

- This PR adds an extra 28KB to repository

cmsbuild commented:
A new Pull Request was created by  @casarsa (Massimo Casarsa) for master.

It involves the following packages:

Validation/MtdValidation


@andrius-k, @kmaeshima, @ErnestaP, @ahmad3213, @cmsbuild, @jfernan2, @rvenditti can you please review it and eventually sign? Thanks.
@fabiocos this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

fabiocos commented on Validation/MtdValidation/plugins/BtlLocalRecoValidation.cc:
```diff
@@ -489,6 +584,52 @@ void BtlLocalRecoValidation::bookHistograms(DQMStore::IBooker& ibook,
+    meCluYRes_ = ibook.book1D("BtlCluYRes", "BTL cluster Y resolution;Y_{RECO}-Y_{SIM} [cm]", 100, -3.1, 3.1);
+    meCluZRes_ = ibook.book1D("BtlCluZRes", "BTL cluster Z resolution;Z_{RECO}-Z_{SIM} [cm]", 100, -0.2, 0.2);
+  }
+  meCluYXLocal_ = ibook.book2D("BtlCluYXLocal",
```
@casarsa don't we want to keep these local position histograms as optional (with the flag)?

casarsa replied:
You are right. I move them into the if satetment

casarsa replied:
and now the size of the MTD histograms is
1135.12 KiB      MTD/ETL
241.00 KiB       MTD/BTL

fabiocos commented on Validation/MtdValidation/plugins/EtlLocalRecoValidation.cc:
```diff
@@ -814,6 +903,92 @@ void EtlLocalRecoValidation::bookHistograms(DQMStore::IBooker& ibook,
+      ibook.book1D("EtlCluZResZneg", "ETL cluster Z resolution (-Z);Z_{RECO}-Z_{SIM} [cm]", 100, -0.003, 0.003);
+  meCluZRes_[1] =
+      ibook.book1D("EtlCluZResZpos", "ETL cluster Z resolution (+Z);Z_{RECO}-Z_{SIM} [cm]", 100, -0.003, 0.003);
+  meCluYXLocal_[0] = ibook.book2D("EtlCluYXLocalZneg",
```
@casarsa same here

casarsa replied:
OK

fabiocos commented:
@casarsa do you have the output plots to attach as documentation? In particular those showing the behaviour studied in the uncertainties tuning?

Do you always manage to find the simhits associated to the cluster? Do you have a check for this?

casarsa committed:
Made the booking and filling of local Y-X histograms optional and added check on REC0-SIM hits matching

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34207/23473

- This PR adds an extra 28KB to repository

cmsbuild commented:
Pull request #34207 was updated. @andrius-k, @kmaeshima, @ErnestaP, @ahmad3213, @cmsbuild, @jfernan2, @rvenditti can you please check and sign again.

casarsa commented:
> Do you always manage to find the simhits associated to the cluster? Do you have a check for this?

I've added a check. From a quick check: out of ~380 BTL reco hits and ~370 reco ETL I always have a match to a sim hit.

fabiocos commented:
please test

@casarsa thanks for checking and the update

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-133f91/16189/summary.html
**COMMIT**: 482a96dbf0c114cb2726e52cbb8c771c067a6b18
**CMSSW**: CMSSW_12_0_X_2021-06-22-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/34207/16189/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 38
 - DQMHistoTests: Total histograms compared: 2785631
 - DQMHistoTests: Total failures: 23
 - DQMHistoTests: Total nulls: 4
 - DQMHistoTests: Total successes: 2785582
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 30.256 KiB( 37 files compared)
 - DQMHistoSizes: changed ( 23234.0,... ): 5.379 KiB        MTD/ETL
 - DQMHistoSizes: changed ( 23234.0,... ): 2.186 KiB        MTD/BTL
 - Checked 160 log files, 37 edm output root files, 38 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-06-22-2300+133f91/43625/triggerResults): no differences found

jfernan2 commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

casarsa commented:
> @casarsa do you understand the pattern of the differences? Why no resolution plot and ETL plot? Didi you see them in the local test output?

Yes, I think I do:
BtlLongposPullvsEta: I changed the SIM hit position in the crystal from mm to cm and this affects the eta.

BtlCluPhi and BtlCluEta: I fixed the calculation of the cluster local position in the module. RectangularMTDTopology::pixelToModuleLocalPoint was used to transform coordinates that already were in the module frame.

BtlTPullvsEta: I've changed the number of bins for uniformity with other similar histograms.

fabiocos commented:
@casarsa thanks, the new histograms are indeed available in the output

qliphy commented:
+1
